File: LeetCode Solutions/May LeetCoding Challenges/Maximum Product of Word Lengths.py
class Solution:
    def maxProduct(self, words: List[str]) -> int:
        words.sort(key=len, reverse=True)
        maximumValue, bitsets = 0, {}
        for i in range(len(words)):
            wlen, bitset = len(words[i]), 0
            if(wlen * len(words[0]) < maximumValue):
                return maximumValue
            for c in words[i]:
                bitset |= 1 << ord(c) - 97
            if(bitset not in bitsets):
                for k,v in bitsets.items():
                    if(not bitset & k):
                        maximumValue = max(maximumValue, wlen * v)
                bitsets[bitset] = wlen
        return maximumValue

File: LeetCode Solutions/May LeetCoding Challenges/Maximum Product of Word Lengths.java
class Solution {
    public int maxProduct(String[] words) {
        int maximumValue = 0;
        int[] bitsets = new int[words.length];
        for(int i=0; i<words.length; i++){
            int wlen = words[i].length(), bitset = 0;
            for(int j=0; j<wlen; j++){
                bitset |= 1 << (words[i].charAt(j) - 'a');
            }
            for(int j=0; j<i; j++){
                if((bitsets[j] & bitset) == 0){
                    maximumValue = Math.max(maximumValue, wlen * words[j].length());
                }
            }
            bitsets[i] = bitset;
        }
        return maximumValue;
    }
}